File: ACM/Acwing/3305/3305.cpp
/*
 * @Author: NEFU AB-IN
 * @Date: 2023-03-04 19:17:20
 * @FilePath: \Acwing\3305\3305.cpp
 * @LastEditTime: 2023-03-04 19:25:45
 */
#include <bits/stdc++.h>
using namespace std;
#define int long long
#undef int

#define SZ(X) ((int)(X).size())
#define ALL(X) (X).begin(), (X).end()
#define IOS                                                                                                            \
    ios::sync_with_stdio(false);                                                                                       \
    cin.tie(nullptr);                                                                                                  \
    cout.tie(nullptr)
#define DEBUG(X) cout << #X << ": " << X << '\n'
typedef pair<int, int> PII;

const int N = 3e4 + 10, INF = 0x3f3f3f3f;

int n, m, k, t;
int a[N];
vector<PII> g[N];
int dist[N], st[N];
int vis[N];

void dij(int s)
{
    priority_queue<PII, vector<PII>, greater<PII>> q;
    q.push({0, s});
    memset(dist, 0x3f, sizeof dist);
    memset(st, 0, sizeof st);
    dist[s] = 0;
    while (SZ(q))
    {
        auto [_, u] = q.top();
        q.pop();
        if (st[u])
            continue;
        st[u] = 1;
        for (auto &[v, w] : g[u])
        {
            if (dist[u] + w <= dist[v])
            {
                dist[v] = dist[u] + w;
                q.push({dist[v], v});
            }
        }
    }
    return;
}

signed main()
{
    IOS;
    cin >> n >> m >> k >> t;
    for (int i = 1; i <= n; ++i)
    {
        cin >> a[i];
    }
    for (int i = 1; i <= m; ++i)
    {
        int x;
        cin >> x;
        g[0].push_back({x, 0});
    }

    for (int i = 1; i <= k; ++i)
    {
        int x, y, z;
        cin >> x >> y >> z;
        int w = max(a[x], a[y]);
        g[x].push_back({z, w});
        g[y].push_back({z, w});
    }
    dij(0);
    cout << dist[t] << '\n';

    return 0;
}

File: ACM/Acwing/3305/3305.py
'''
Author: NEFU AB-IN
Date: 2023-03-19 11:26:46
FilePath: \Acwing\3305\3305.py
LastEditTime: 2023-03-19 11:42:51
'''
read = lambda: map(int, input().split())
from collections import Counter, deque
from heapq import heappop, heappush
from itertools import permutations

N = int(2e3 + 10)
INF = int(2e9)
st, w, dist = [0] * N, [0] * N, [INF] * N
g = [[] for _ in range(N)]
q = deque()


def spfa():
    while len(q):
        u = q.pop()
        st[u] = 0
        for v, c in g[u]:
            if max(dist[u], dist[v]) + max(w[u], w[v]) < dist[c]:
                dist[c] = max(dist[u], dist[v]) + max(w[u], w[v])
                if st[c] == 0:
                    st[c] = 1
                    q.appendleft(c)


n, m, k, t = read()
w[1:] = list(read())

tmp = list(read())
for i in tmp:
    q.appendleft(i)
    st[i] = 1
    dist[i] = 0

for i in range(k):
    a, b, c = read()
    g[a].append([b, c])
    g[b].append([a, c])

spfa()

print(dist[t])
